File: TypeScript/event.ts
/**
 * @Author: tangchengqin
 * @Date: 2024/9/14 10:25:16
 * @LastEditors: tangchengqin
 * @LastEditTime: 2024/9/18 16:08:34
 * Description: 
 * Copyright: Copyright (©)}) 2024 Clarify. All rights reserved.
 */

export interface Params {
    [param: string]: any;
}

export class Event {

    public eventName: string;
    private _evtId: number;
    private _callback: Function;
    private _filter?: Function;

    constructor(evtId: number, eventName: string, callback: Function, filter?: Function) {
        this._evtId = evtId;
        this.eventName = eventName;
        this._callback = callback;
        this._filter = filter;
    }

    public get evtId(): number {
        return this._evtId;
    }

    public execute(args: Params): void {
        if (this._filter && !this._filter(args)) {
            return;
        }
        this._callback(args);
    }
}

File: Go/event.go
/**
 * @Author: tangchengqin
 * @Date: 2024/9/18 11:47:49
 * @LastEditors: tangchengqin
 * @LastEditTime: 2024/9/18 16:27:23
 * Description: 
 * Copyright: Copyright (©)}) 2024 Clarify. All rights reserved.
 */
package event

type Event struct {
	evtId int
	eventName string
	callback func(...any)
	filter func(...any) bool
}

func (slf *Event) GetID() int {
	return slf.evtId
}

func (slf *Event) GetName() string {
	return slf.eventName
}

func (slf *Event) Execute(args ...any) {
	if slf.filter != nil && !slf.filter(args) {
		return
	}
	slf.callback(args)
}
